src/WebJobs.Script/Description/PowerShell/PowerShellFunctionInvoker.cs

```diff
@@ -0,0 +1,450 @@
+
+        private readonly Collection<FunctionBinding> _inputBindings;
+        private readonly Collection<FunctionBinding> _outputBindings;
+        private readonly IMetricsLogger _metrics;
+
+        private string _script;
+        private List<string> _moduleFiles;
+        private Dictionary<string, string> _environmentVariables;
+
+        internal PowerShellFunctionInvoker(ScriptHost host, FunctionMetadata functionMetadata,
+            Collection<FunctionBinding> inputBindings, Collection<FunctionBinding> outputBindings) : base(host, functionMetadata)
+        {
+            _host = host;
+            _scriptFilePath = functionMetadata.ScriptFile;
+            _functionName = functionMetadata.Name;
+            _inputBindings = inputBindings;
+            _outputBindings = outputBindings;
+            _metrics = host.ScriptConfig.HostConfig.GetService<IMetricsLogger>();
+            _environmentVariables = new Dictionary<string, string>();
+        }
+
+        public override async Task Invoke(object[] parameters)
+        {
+            // TODO: Refactor common code for providers.
+            object input = parameters[0];
+            TraceWriter traceWriter = (TraceWriter)parameters[1];
+            IBinderEx binder = (IBinderEx)parameters[2];
+            ExecutionContext functionExecutionContext = (ExecutionContext)parameters[3];
+            string invocationId = functionExecutionContext.InvocationId.ToString();
+
+            FunctionStartedEvent startedEvent = new FunctionStartedEvent(functionExecutionContext.InvocationId, Metadata);
+            _metrics.BeginEvent(startedEvent);
+
+            try
+            {
+                object convertedInput = input;
+                if (input != null)
+                {
+                    HttpRequestMessage request = input as HttpRequestMessage;
+                    if (request != null)
+                    {
+                        // TODO: Handle other content types? (E.g. byte[])
+                        if (request.Content != null && request.Content.Headers.ContentLength > 0)
+                        {
+                            convertedInput = ((HttpRequestMessage)input).Content.ReadAsStringAsync().Result;
+                        }
+                    }
+                }
+
+                TraceWriter.Info(string.Format("Function started (Id={0})", invocationId));
+
+                string functionInstanceOutputPath = Path.Combine(Path.GetTempPath(), "Functions", "Binding",
+                    invocationId);
+
+                Dictionary<string, string> bindingData = GetBindingData(convertedInput, binder);
+                bindingData["InvocationId"] = invocationId;
+
+                await
+                    ProcessInputBindingsAsync(convertedInput, functionInstanceOutputPath, binder, bindingData,
+                        _environmentVariables);
+
+                InitializeEnvironmentVariables(_environmentVariables, functionInstanceOutputPath, input,
```

Comment:
Shouldn't the environment variables here be local? We don't want to share this state between function invocations.